src/compiler.ts

```diff
@@ -4566,13 +4564,15 @@ export class Compiler extends DiagnosticEmitter {
               expr = module.if(leftExpr, rightExpr, module.i32(0));
             }
           }
+          flow.inheritBranch(rightFlow, condKind);
           this.currentFlow = flow;
           this.currentType = Type.bool;
 
         } else {
           rightExpr = this.compileExpression(right, leftType, inheritedConstraints | Constraints.CONV_IMPLICIT);
           rightType = this.currentType;
           rightFlow.freeScopedLocals();
+          flow.inheritBranch(rightFlow);
```

Comment:
Let's move this line one up, so it happens before `freeScopedLocals` just in case, that is, avoiding undesired side-effects from freeing.